### ISSUE TITLE
Update broke traditionalBuilder - key files not generated

### ISSUE THREAD

**Issue by sethblumberg** [SublimeText/LaTeXTools#606]
Since the latest update (11/27/15), LaTeXTools fails to typset under traditionalBuilder.  The pdf file isn't produced, or if exists isn't updated.

pdflatex (with no flags or modifiers) works fine from the terminal, so underlying LaTeX is okay.

My setup: Sublime 2.0.2, newest LaTeXTools (files modified 11-27-15, though readme says 11-22-15), OSX 10.9.5 Mavericks, TexLive-2014

I tried to build a simple sample file, whose entire contents are:

```
\documentclass{article} \title{Title} \author{Your Name} \begin{document} \maketitle{} \section{Introduction} This is where you will write your content. \end{document}
```

The build message suggests all is well:
```
[Compiling /Users/seth/Desktop/sample/sample.tex]
TraditionalBuilder: Engine: pdflatex -> pdf. Invoking latexmk... done.
```

But the project folder file only gets two additional files: `sample.aux` and `sample.fdb_latexmk`.

The console reflects that the other files weren't created, but otherwise doesn't seem too helpful.  Here's what it shows:

```
<module 'traditionalBuilder' from '/Users/seth/Library/Application Support/Sublime Text 2/Packages/LaTeXTools/builders/traditionalBuilder.pyc'>
<class 'traditionalBuilder.TraditionalBuilder'>
Welcome to thread Thread-3
2
[u'latexmk', u'-cd', u'-e', u"$pdflatex = 'pdf -enable-write18 -interaction=nonstopmode -synctex=1 %S %O'", u'-f', u'-pdf', u'sample.tex']
Finished normally
12
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.6/lib/python2.6/threading.py", line 532, in __bootstrap_inner
    self.run()
  File "./makePDF.py", line 164, in run
IOError: [Errno 2] No such file or directory: u'/Users/seth/Desktop/sample/sample.log'
```

Is there an open issue about this already?  It seems strange that the update broke compilation but no one else posted about it.

**Comment by Limro** [comment]
Can you try running it with [Sublime Text 3](http://www.sublimetext.com/3)?
It could be an issue with ST2.

**Comment by r-stein** [comment]
Can you open a shell in the folder and try to run:
```sh
latexmk -cd -e $pdflatex='pdf -enable-write18 -interaction=nonstopmode -synctex=1 %S %O' -f -pdf sample.tex
```

**Comment by ig0774** [comment]
You're right this is behaviour introduced in the latest build by the changes required to make the `!TEX option =` parsing work. It hasn't been reported because it only breaks the compile functionality if you've customised the command in your `LaTeXTools.sublime-settings` file, which you appear to have done to ensure the `-enable-write18` flag is set, and only if you are using pdflatex.

In order to restore functionality _for now_ I recommend _removing_ your custom command entry and instead, add an `options` setting to your `LaTeXTools.sublime-settings` file like so:

```json
"options": [
    "-enable-write18"
]
```

I'll look into what can be done to fix this when I get to a computer.

For anyone whose interested, the underlying problem is that explicitly setting the `$pdflatex` variable stops latexmk from modifying the underlying command, so in the last update, I switched to using latexmk's "-pdf" switch to ensure pdflatex is set and that it will accept the "-latex-option" parameters. This changes the value subbed for "%E" in the command to `pdf`, hence the issue (see the generated value of `$pdflatex` which is now trying to run an executable called `pdf`)

**Comment by msiniscalchi** [comment]
I think the right approach here is to first test for a custom command; if
so, all the option / engine parsing logic should be skipped. I'll also look
into this. I was hoping to do a new release this weekend, but Thanksgiving
happened :) Hopefully, later in the week.

On Mon, Nov 30, 2015 at 5:15 AM, ig0774 <notifications@github.com> wrote:

> You're right this is behaviour introduced in the latest build by the
> changes required to make the !TEX option = parsing work. It hasn't been
> reported because it only breaks the compile functionality if you've
> customised the command in your LaTeXTools.sublime-settings file, which
> you appear to have done to ensure the -enable-write18 flag is set, and
> only if you are using pdflatex.
>
> In order to restore functionality *for now* I recommend *removing* your
> custom command entry and instead, add an options setting to your
> LaTeXTools.sublime-settings file like so:
>
> "options": [
>     "-enable-write18"
> ]
>
> I'll look into what can be done to fix this when I get to a computer.
>
> For anyone whose interested, the underlying problem is that explicitly
> setting the $pdflatex variable stops latexmk from modifying the
> underlying command, so in the last update, I switched to using latexmk's
> "-pdf" switch to ensure pdflatex is set and that it will accept the
> "-latex-option" parameters. This changes the value subbed for "%E" in the
> command to pdf, hence the issue (see the generated value of $pdflatex
> which is now trying to run an executable called pdf)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/SublimeText/LaTeXTools/issues/606#issuecomment-160601472>
> .
>



-- 
Marciano Siniscalchi
Economics Department, Northwestern University
http://faculty.wcas.northwestern.edu/~msi661

**Comment by ig0774** [comment]
#607 should address this issue and continue to allow custom `$pdflatex =` settings, such as the one @sethblumberg is using. I should add an update to the documentation for TeX options, however, as it does not work if a custom `$pdflatex` command is specified (this is a limitation of latexmk).

**Comment by sethblumberg** [comment]
Thanks everyone.  Actually, I had customized the `traditionalBuilder.py` file - I hadn't realized how to activate shell escape from within `LaTeXTools.sublime-settings`.  I needed shell escape (write-18) for the `auto-pst-pdf` option of `pstricks`.

I replaced my builder file with the default one, and I added the write18 option to my settings file as @ig0774 suggested (under `"builder_settings"`).  Now everything runs fine.

One related question: The readme says that options can be set at the top of the `.tex` file.  But what is the syntax to pass write18?  The obvious thing to try was

```
%TEX options = "-enable-write18"
```

but neither that nor any variants worked.  And is this standard LaTeX, or just for LaTeXTools?

**Comment by ig0774** [comment]
It should actually be:

```
%!TEX option = -enable-write18
```

Note the `!`.

I think the syntax may be supported somewhere else, but it's implemented in the editor rather than by LaTeX itself.

**Comment by sethblumberg** [comment]
That works (!).  The readme needs the "!".

**Comment by sverley** [comment]
Hi,

I have exactly the same problem but I work on ST3 on linux with a texlive 2010.

Can you help me to solve this issue ?

**Comment by ig0774** [comment]
@sverley: You actually have a completely different problem (the one here was fixed in #607). 

To support the `%!TEX options` stuff, I migrated to a default command that actually only works on latexmk v4.31 or later (v4.37 or later if you're using `lualatex`), which is probably not available on TeXLive 2010 (v4.31 came out in 2012). You can check which version you have by running `latexmk --version`. 

AFAICT you have a few of options (basically in order of recommendation):
 1. Download and install a newer version of `latexmk` [from the author's site](http://users.phys.psu.edu/~collins/software/latexmk/). You'll have to make the Perl script executable, rename it from `latexmk.pl` `latexmk`, and put it somewhere on your path that's found before your TeX executables, but otherwise this should work.
 1. Override the default command in your LaTeXTools.sublime-settings file. To do this find the `"builder_settings"` dictionary and add an entry like this:

 ```json
 "command": ["latexmk", "-cd", "-e", "$pdflatex = '%E -interaction=nonstopmode -synctex=1 %S %O'", "-f", "-pdf"],
 ```

 Which should return things to using the old version of the command, though do pay attention to the various caveats given for modifying this setting in the README. Among other things, this configuration won't support the `%!TEX option` parsing.

Much more destructively (and I don't necessarily recommend these) you could:
 1. Update TeXLive, but I'm assuming you have a reason to stay with TeXLive 2010.
 1. Using the (new) releases option to roll-back to v3.4.5 (before this change was made) and staying there (using the `auto_upgrade_ignore` setting of Package Control). Obviously this means you miss out on any new features that are added, bugs fixed, etc..

**Comment by sverley** [comment]
Thanks for the answer. Actually, I tried with the texlive2013 and it's work but I cannot update it permanently since I am not root. 
Anyway, thanks for the two options you suggest works perfertly. I read the ticket you mentionned, but I didn't understand what to do... My fault :-)
